test/e2e/predicates.go

```diff
@@ -126,30 +144,26 @@ var _ = Describe("Predicates E2E Test", func() {
 			},
 		}
 
-		job := &jobSpec{
-			name: "pa-job",
+		job1 := &jobSpec{
+			name: "pa-job1",
 			tasks: []taskSpec{
 				{
 					img:      "nginx",
-					req:      slot,
-					min:      rep,
-					rep:      rep,
+					req:      halfCPU,
+					min:      1,
+					rep:      1,
 					affinity: affinity,
-					labels:   labels,
 				},
 			},
 		}
 
-		_, pg := createJobEx(context, job)
-		err := waitPodGroupReady(context, pg)
+		_, pg1 := createJobEx(context, job1)
+		err = waitPodGroupReady(context, pg1)
 		Expect(err).NotTo(HaveOccurred())
 
-		pods := getPodOfPodGroup(context, pg)
-		// All pods should be scheduled to the same node.
-		nodeName := pods[0].Spec.NodeName
-		for _, pod := range pods {
-			Expect(pod.Spec.NodeName).To(Equal(nodeName))
-		}
+		podsWithAffinity := getPodOfPodGroup(context, pg1)
```

Comment:
why do we change that?

Reply:
All the pods are getting unschedulable, so all the pods are in pending state, so it is getting time out while checking `waitPodGroupReady`, so have changed the test case such that, first a task with one pod and label is deployed, and then second task with podAffinity is scheduled, and checking whether both are having same node name, verifying pod affinity predicate is working.

Reply:
![screenshot from 2019-02-20 16-18-26](https://user-images.githubusercontent.com/34565833/53087162-80f6c580-352c-11e9-902b-260076a42323.png)
This is resource of DIND-nodes, that is being created in travis, one node always has difference in resources, so if is it trying to schedule pods in that node, for the `rep` value of other nodes, then one pod goes unschedulable, so `waitPodGroupReady` returns error, so test case fails, to avoid that case, this case where one pod is created with label and next pod with `podAffinity` and check is done whether both pods are same scheduled in same node or not, that verifies our pod affinity predicate avoid unschedulable scenario.

Reply:
if so, just  change oneCpu to halfCpu when creating the Job should be enough.

Reply:
In InterPodAffinity Priority function, they try to get the node of existingPod from it's Pod object, 
```
processPod := func(existingPod *v1.Pod) error {
		existingPodNode, err := ipa.info.GetNodeInfo(existingPod.Spec.NodeName)
                ........
                ........
    }
```
But in gang scheduling, we bind task only when the number of tasks is greater or equal to the MinAvailable number.
```
        occupied := readyTaskNum(job)
	return occupied >= job.MinAvailable
```
So if our job has two tasks, and first get allotted to a node and the cache(schedulercache) info in our scheduler gets reflected with the new pod(v1.Pod) Object, but that pod object will not have nodeName(pod.Spec.NodeName), so when second task comes and checks for InterPodAffinity score, it checks for all the pods from the cache(schedulercache.NodeInfo), and tries to find the nodeName of the pod from Pod object(v1.Pod), but for first task in our case will not have nodeName in pod(v1.Pod) object, since it is not yet sent for binding, it will be in pending state, so for second task, we will get error that no node with nill value(empty string)

`Calculate Inter Pod Affinity Priority Failed because of Error: failed to find node <>`

So both not being able to switch from Pending to Running state since the `occupied value will be less than MinAvailable`, so our gang scheduling will not work with K8s InterPod Affinity logic.
In K8s Pod Affinity Predicates, they use `topologyPairsAntiAffinityPodsMap` in their `predicateMetaData`, and run predicates for the task, so predicates was working fine with gang scheduling, but InterPodAffinity Priority function in K8s will not support gang scheduling model.

Reply:
If so, we need to find a solution. Affiinity to itself is a valid case.

Reply:
Maybe we can deepcopy pods and set its nodename by TaskInfo when listing. BTW, only deep copy the pod that NodeName is empty but TaskInfo.NodeName is NOT empty.

Reply:
But that listing is done by InterPod Affinity code provided by K8s, if we have to change the code, we will have to bring it out of vendor and place it in kube-batch.  We just send current pod object(v1.Pod), list of nodes(both schedulercache.NodeInfo(as map) and v1.Nodes(as slice)).

Reply:
> But that listing is done by InterPod Affinity code provided by K8s

It will use lister from our code, right?

Reply:
Yes, have made change in our lister code, and Affinity to itself case also passes now.  And have changed predicate Pod affinity test case to the one that was existing before.

Reply:
LGTM overall, please only do DeepCopy when `task.NodeName != pod.NodeName`.

Reply:
We don't use `List` and `FilteredList` function in interPodAffinity priority code.  Even if that check is included, it will not affect behaviour of priority, so have just implemented the same as that of implemented in predicate plugin.